### ISSUE TITLE
NativeAot isinst branch trimming

### ISSUE THREAD

**Issue by NinoFloris** [dotnet/runtime#92850]
NativeAot isinst branches should be trimmed.

Given the following code:

```cs
using System.Runtime.CompilerServices;
using System.Threading;
using System.Threading.Tasks;

var intFoo = new Foo<int>(new Base<int>());
var floatFoo = new Foo<float>(new Base<float>());
// ... and other value type instantiations of Foo.

await intFoo.GetAsync();
await floatFoo.GetAsync();

class Base<T>
{
    public virtual T Get() => default!;
}

class AsyncBase<T>: Base<T>
{
    public virtual async ValueTask<T> GetAsync(CancellationToken cancellationToken = default)
    {
        await Task.Yield(); // pseudocode
        return default!;
    }
}

class Foo<T>
{
    readonly Base<T> _instance;
    public Foo(Base<T> instance) => _instance = instance;

    // here for easier mstat diag. 
    [MethodImpl(MethodImplOptions.NoInlining)]
    public Task<T> GetAsync(CancellationToken cancellationToken = default)
    {
        if (_instance is AsyncBase<T> asyncBase)
            return asyncBase.GetAsync(cancellationToken).AsTask();

        return Task.FromResult(_instance.Get());
    }
}
```

<img width="1318" alt="Screenshot 2023-09-30 at 16 56 20" src="https://github.com/dotnet/runtime/assets/4218809/4afaebe7-ac43-489f-93ec-1b4c9012e4d8">

We notice a deficiency around the instance of check. Even though it would be *impossible* for this code to call the `GetAsync` branch on any `Base<T>` instance it'll keep the unreachable ValueTaskSourceAsTask instantiations in the binary. Could the branch contents be removed?

I know the trimmer doesn't do much control flow analysis by design (though it is able to drop the metadata and code for `AsyncBase<int>.GetAsync`, so it's not entirely absent) but this deficiency severely limits our ability to optimize bloat based on type capabilities.

cc @MichalStrehovsky

**Comment by MichalStrehovsky** [comment]
TL;DR; there's a workaround at the bottom.

The `isinst` branch is already trimmed. Here's the assembly code:

```
    31:     // here for easier mstat diag. 
    32:     [MethodImpl(MethodImplOptions.NoInlining)]
    33:     public Task<T> GetAsync(CancellationToken cancellationToken = default)
    34:     {
    35:         if (_instance is AsyncBase<T> asyncBase)
00007FF626ACBE80  sub         rsp,28h  
00007FF626ACBE84  mov         rax,qword ptr [rcx+8]  
    36:             return asyncBase.GetAsync(cancellationToken).AsTask();
    37: 
    38:         return Task.FromResult(_instance.Get());
00007FF626ACBE88  cmp         byte ptr [rax],al  
00007FF626ACBE8A  lea         rax,[S_P_CoreLib_System_Threading_Tasks_TaskCache::__NONGCSTATICS (07FF626AFEA10h)]  
00007FF626ACBE91  cmp         qword ptr [rax-8],0  
00007FF626ACBE96  jne         asyncopt_Foo_1<Int32>__GetAsync+32h (07FF626ACBEB2h)  
00007FF626ACBE98  mov         rax,qword ptr [S_P_CoreLib_System_Threading_Tasks_TaskCache::__GCSTATICS (07FF626B95408h)]  
00007FF626ACBE9F  mov         rax,qword ptr [rax+18h]  
00007FF626ACBEA3  cmp         dword ptr [rax+8],1  
00007FF626ACBEA7  jbe         asyncopt_Foo_1<Int32>__GetAsync+39h (07FF626ACBEB9h)  
00007FF626ACBEA9  mov         rax,qword ptr [rax+18h]  
00007FF626ACBEAD  add         rsp,28h  
00007FF626ACBEB1  ret  
00007FF626ACBEB2  call        __GetGCStaticBase_S_P_CoreLib_System_Threading_Tasks_TaskCache (07FF626A21447h)  
00007FF626ACBEB7  jmp         asyncopt_Foo_1<Int32>__GetAsync+18h (07FF626ACBE98h)  
00007FF626ACBEB9  call        S_P_CoreLib_Internal_Runtime_CompilerHelpers_ThrowHelpers__ThrowIndexOutOfRangeException (07FF626ABC000h)  
00007FF626ACBEBE  int         3  
```

(You're looking at an inlined `Get` and inlined `Task.FromResult`.)

We can't answer whether an isinst branch would be taken without doing whole program analysis first. Whole program analysis operates on the level of methods, not basic blocks (i.e. we scan the `AsTask()` part too when building the whole program view). We have optimizations that limit graph expansion but they don't work well on valuetypes (this is calling a method on a value task). They also don't work well if the whole program view figured out something is a target of reflection.

What happens here is that somewhere in `AsTask`, the code creates a delegate to `g_cctorBlahBlah` method you have highlighted. Once that happens during whole program view building, we can no longer get rid of it.

Sizoscope is not telling you the whole picture. You can see that this happened if you delete the `*.scan.dgml.xml` next to the mstat and rename the `*.codegen.dgml.xml` to `*scan.dgml.xml` (Sizoscope only opens the scan one). If you reopen the mstat, Sizoscope will say the method is kept because of reflection, not because of the branch not being removed. The codegen DGML file the the one from codegen, the scan DGML is from the whole program analysis phase.

Breaking up whole program analysis to operate on pieces of methods (basic blocks) so that we can delay visiting the branch is something we might consider if we ever do #83021. I'll add a note to the issue. However, this is the [halting problem](https://en.wikipedia.org/wiki/Halting_problem) in general and there's only so much effort we're willing to spend to partially solve it.

If you want to work around, you need to play into the optimization strengths around removing methods on classes. Add this method to `AsyncBase<T>`:

```csharp
    public Task<T> GetAsyncTask(CancellationToken cancellationToken = default)
    {
        return GetAsync(cancellationToken).AsTask();
    }
```

And call it instead of the `return asyncBase.GetAsync(cancellationToken).AsTask();`.

**Comment by NinoFloris** [comment]
Interesting. Thanks a lot for the detailed explanation! 

Whenever you have a moment could you take a look at why the following fails to trim?
Your workaround seems to only work if the method is on AsyncBase itself.

```cs
using System.Runtime.CompilerServices;
using System.Threading;
using System.Threading.Tasks;

var intFoo = new Foo<int>(new Base<int>());
var floatFoo = new Foo<float>(new Base<float>());
// ... and other value type instantiations of Foo.

await intFoo.GetAsync();
await floatFoo.GetAsync();

class Base<T>
{
    public virtual T Get() => default!;
}

class AsyncBase<T>: Base<T>
{
    public virtual async ValueTask<T> GetAsync(CancellationToken cancellationToken = default)
    {
        await Task.Yield(); // pseudocode
        return default!;
    }
}

class Foo<T>
{
    readonly Base<T> _instance;
    public Foo(Base<T> instance) => _instance = instance;

    // here for easier mstat diag.
    [MethodImpl(MethodImplOptions.NoInlining)]
    public Task<T> GetAsync(CancellationToken cancellationToken = default)
    {
        if (_instance is AsyncBase<T> asyncBase)
            return GetAsyncTask(asyncBase, cancellationToken);

        return Task.FromResult(_instance.Get());
    }

    static Task<T> GetAsyncTask(AsyncBase<T> instance, CancellationToken cancellationToken = default)
    {
        return instance.GetAsync(cancellationToken).AsTask();
    }
}
```

**Comment by MichalStrehovsky** [comment]
We delay scanning instance methods when building whole program view until we see an allocation. Placing the instance method on something that is allocated in the program defeats the purpose.

**Comment by NinoFloris** [comment]
Alright, that makes sense.

For a moment there I assumed, considering you mentioned:

> If you want to work around, you need to play into the optimization strengths around **removing methods on classes**.

It would just work as long as the method would only be reached from the branch. But I can see how that would basically run into the same basic block analysis requirement again, just with an additional hop.
